### ISSUE TITLE
[prefer-optional-chain] Prefer optional chain recommendation changes type when accessing a boolean

### ISSUE THREAD

**Issue by Calme1709** [typescript-eslint/typescript-eslint#2963]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
	"parser": "@typescript-eslint/parser",
	"plugins": [
		"@typescript-eslint",
	],
	"rules": {
		"@typescript-eslint/prefer-optional-chain": "error"
	}
}
```

```TS
interface MyInterface {
	array: string[];
}

let myObject: MyInterface | undefined;

const objectIncludesTest: boolean = myObject !== undefined && myObject.array.includes("test");
```

A minimal repo for reproduction can be found [here](https://github.com/Calme1709/Typescript-Eslint-Bug)

**Expected Result**
In this case I would expect that no error would be thrown.

**Actual Result**
An error is raised suggesting that:

```TS
const objectIncludesTest = myObject !== undefined && myObject.array.includes("test");
```

should become

```TS
const objectIncludesTest = myObject?.array.includes("test");
```

However if this is done the type of objectIncludesTest is now of type ```boolean | undefined``` instead of ```boolean``` as it was before.

**Additional Info**
It would make more sense if an error was not raised if it would change the type of the expression.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.0` |
| `@typescript-eslint/parser`        | `4.14.0` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.18.0` |
| `node`                             | `15.5.0` |

**Comment by bradzacher** [comment]
The rule uses a suggestion fixer for this reason.
Suggestion fixers are not necessarily safe to apply and it's up to the engineer to assess them and determine if they should be applied.

This is why suggestion fixers cannot be applied automatically - only manually via the IDE.

Detecting types would require type information, which would make the rule too heavy.
